### ISSUE TITLE
[FR] Better control over Send Message process & not loosing data

### ISSUE THREAD

**Issue by henry2man** [flyerhq/flutter_chat_ui#290]
## Is your feature request related to a problem?

Imagine that a user composes a long message, press send... and there isn't any network connectivity. Currently the message field is cleared always, regardless of whether the message got delivered or not. You can show some kind of error notification to the user, but the message is lost.

## What solution would you suggest?

`Input.onSendPressed` method signature should be changed to `FutureOr<bool> Function(types.PartialText) onSendPressed` (or simply `Future<bool> Function(types.PartialText) onSendPressed`), so any send function must return `true` or `false`, indicating that the message was (or wasn't) successfully delivered.

Also TextField must react depending on onSendPressed result, **only clearing data if the message was successfully delivered** and maintaining the data if wasn't.

## Is there any additional solution to that?
n/a

## Extras
**Screenshots or videos 📸**
n/a

**Code snippets 📝**
n/a

## Related issues/PRs
n/a

**Comment by maximveksler** [comment]
Most Def. A critical component of production grade chat experience.

**Comment by demchenkoalex** [comment]
thanks, will think how to address this! not sure I want people to return true or false, because most definitely most will forget to do so, but yeah, something to think about.

**Comment by henry2man** [comment]
Well, with this (breaking) change all devs must will need to update code, so the new behaviour will be noticed, for sure. And the framework needs to know if the sending process is completed or not. 

You may add a "`onSendError`" callback but this will not prevent textBox text deletion.  

For now I'm not able to see another alternative but, of course, I'm open for other alternatives if we can increase the robustness

**Comment by demchenkoalex** [comment]
Just started thinking about it and actually - is it needed at all? I went to the Firebase repo to check the new API and realised message is always sent (Firebase manages merging of local database with remote one when internet is available) and I thought, how it will look like in messengers? Turns out that with the airplane mode, all, Telegram, Messenger, WhatsApp and Signal always send message and clear input. Then everything is managed on the message itself (and we have sending and error states there). So adding this will encourage people using low quality messages statuses? I would prefer having a sync API like we have now, but maybe give this option hidden somewhere, just in case. Maybe we should just export a method to clear textfield programmatically and add a textfield clear behaviour, with the default `always` (like now) and `never` - developer will need to clear it automatically if he does async stuff in `onSendPressed`. That works?

**Comment by demchenkoalex** [comment]
Did it this way, I think it will do for now, to implement this better control, add to `Chat` widget:

```dart
inputOptions: InputOptions(
  inputClearMode: InputClearMode.never,
  textEditingController: _controller,
),
```

declare the controller (note I am using our custom controller imported from the library, to have the markdown highlight in the input, but this can be a simple `TextEditingController` as well)

```dart
final _controller = InputTextFieldController();
```

and then somewhere in `onSendPressed` - 

```dart
_controller.clear();
```

**Comment by demchenkoalex** [comment]
Will be included in `v1.6.4`

**Comment by henry2man** [comment]
> So adding this will encourage people using low quality APIs and messages statuses handling? 

Well, sometimes you're a solo dev and you have your handmade backend  so it's not always possible something comparable to WhatsApp or Telegram 😅

So I don't think this use cases implies "low quality APIs"; just custom ones.

Will give it a try with `1.6.4`, thanks!!
